routes/bridges/gassistant/index.js

```diff
@@ -0,0 +1,235 @@
+        }));
+    }
+
+    sendChoice(idx, what, title, text) {
+        let suggestions = [];
+        // Filter out buttons more than 25 characters long
+        // since Google Assistant has a cap of 25 characters
+        if (title.length <= 25)
+            suggestions.push(title.substring(0, 25));
+        else
+            console.log(`${title} exceeds max length of 25 characters`);
+        if (suggestions.length)
+            this._buffer.push(new Suggestions(suggestions));
+    }
+
+    sendButton(title, json) {
+        let suggestions = [];
+        // Filter out buttons more than 25 characters long
+        // since Google Assistant has a cap of 25 characters
+        if (title.length <= 25)
+            suggestions.push(title.substring(0, 25));
+        else
+            console.log(`${title} exceeds max length of 25 characters`);
+        if (suggestions.length)
+            this._buffer.push(new Suggestions(suggestions));
+    }
+
+    sendLink(title, url) {
+        if (url === '/user/register') {
+            this._requestSignin = true;
+        } else {
+            this._buffer.push(new Button({
+                title: title,
+                url: url,
+            }));
+        }
+    }
+
+    sendResult(message, icon) {
+        if (typeof this._buffer[this._buffer.length - 1] === 'string')
+            // If there is already a text reply immediately before, we merge text replies
+            // because Google Assistant limits at most 2 chat bubbles per turn
+            this._buffer[this._buffer.length - 1] += '\n' + message.toLocaleString(this._locale);
+        else
+            this._buffer.push(message.toLocaleString(this._locale));
+    }
+
+    sendAskSpecial(what) {
+        // TODO
+    }
+}
+GoogleAssistantDelegate.prototype.$rpcMethods = ['send', 'sendPicture', 'sendChoice', 'sendLink', 'sendButton', 'sendAskSpecial', 'sendRDL', 'sendResult'];
+
+function authenticate(req, res, next) {
+    console.log(req.body.user);
+    if (req.body.user.accessToken) {
+        req.headers.authorization = 'Bearer ' + req.body.user.accessToken;
+        passport.authenticate('bearer', { session: false })(req, res, next);
+    } else {
+        next();
+    }
+}
+
+router.use(authenticate);
+router.use(userUtils.requireScope('user-exec-command'));
+
+const app = actionssdk();
+
+async function retrieveUser(accessToken) {
+    let anonymous, user;
+    if (accessToken) {
+        const decoded = await util.promisify(jwt.verify)(accessToken, secret.getJWTSigningKey(), {
+            algorithms: ['HS256'],
+            audience: 'oauth2',
+            clockTolerance: 30,
+        });
+        user = await db.withClient(async (dbClient) => {
+            const rows = await model.getByCloudId(dbClient, decoded.sub);
+            if (rows.length < 1) {
+                anonymous = true;
+                return await userUtils.getAnonymousUser();
+            }
+            await model.recordLogin(dbClient, rows[0].id);
```

Comment:
You've already authenticated the user, you don't need to recordLogin again.

Reply:
Fixed